[SymbolGraphGen] Correctly handle exported imports in swift-symbolgraph-extract

daniel-grumberg opened the pull request:
This sets up the necessary stuff to include symbols from exported imported modules in symbol graphs generated using the `swift-symbolgraph-extract` tool

Changes include:
- Add option to provide an allow list of reexported modules for swift-symbolgraph-extract
- Recursively collect exported imports to allow fetching all visible Decls for symbol graph generation
- Enable collection of exported imports that originate from non `SourceFile` file units for the purpose of `ModuleDecl::getDisplayDecl`

Resolves #59920 
rdar://89687175

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

daniel-grumberg committed:
Add option to provide an allow list of reexported modules for swift-symbolgraph-extract

daniel-grumberg committed:
Recursively collect exported imports to allow fetching all visible Decls for symbol graph generation

This change is two fold. Firstly it enables collection of exported
imports from non source file units. Additionally this recurses through
the exported imports to ensure the transitive set is collected.

Fixes https://github.com/apple/swift/issues/59920
rdar://89687175

daniel-grumberg commented:
@swift-ci please test

daniel-grumberg commented:
@swift-ci please smoke test

daniel-grumberg commented:
@swift-ci please test

daniel-grumberg commented:
@swift-ci please test

daniel-grumberg commented on lib/AST/Module.cpp:
```diff
@@ -1543,40 +1509,100 @@ SourceFile::getExternalRawLocsForDecl(const Decl *D) const {
+    this->getDisplayDeclsRecursivelyAndImports(Results, importCollector);
+  }
   // FIXME: Should this do extra access control filtering?
   FORWARD(getDisplayDecls, (Results));
```
Make this an else branch

QuietMisdreavus reviewed (approved):
Other than the comment we spotted while walking through the code, could you add a test that verifies that the recursive export-import functionality still works? There's a similar test in `SymbolGraph/Module/ThirdOrder.swift` that you can copy to make something similar. We just need to make sure that the innermost module's symbols appear in the outer symbol graph.

daniel-grumberg committed:
Add test to check nested exported imports in SymbolGraphGen

daniel-grumberg committed:
Fix `ModuleDecl::getDisplayDecls` for recursive case.

Ensure recursive calls to `ModuleDecl::getDisplayDecls` does not result in double collection of display decls.

daniel-grumberg commented:
@swift-ci please test

daniel-grumberg commented:
> Other than the comment we spotted while walking through the code, could you add a test that verifies that the recursive export-import functionality still works? There's a similar test in `SymbolGraph/Module/ThirdOrder.swift` that you can copy to make something similar. We just need to make sure that the innermost module's symbols appear in the outer symbol graph.

Test is in `test/SymbolGraph/Module/NestedExportedImport.swift`

QuietMisdreavus reviewed (approved):
New test looks great, thanks!
